text/0004-remove-unnecessary-allocator-usage.md

```diff
@@ -0,0 +1,212 @@
+
+Host implementaters should be aware that, because a non-zero identifier value was previously valid, this might require slightly more changes to the host than just adding the new function.
+
+```wat
+(func $ext_offchain_http_request_write_body_version_2
+  (param $method i64) (param $uri i64) (param $meta i64) (result i32))
+```
+
+The behaviour of this function is identical to its version 1 equivalent. Instead of allocating a buffer, writing two bytes in it, and returning a pointer to it, the new version of this function simply indicates what happened:
+
+- 0 on success.
+- 1 if the deadline was reached.
+- 2 if there was an I/O error while processing the request.
+- 3 if the identifier of the request is invalid.
+
+These values are equal to the values returned on error by the version 1 (see <https://spec.polkadot.network/chap-host-api#defn-http-error>), but incremented by 1 in order to reserve 0 for success.
+
+```wat
+(func $ext_offchain_submit_transaction_version_2
+    (param $data i64) (return i32))
+(func $ext_offchain_http_request_add_header_version_2
+    (param $request_id i32) (param $name i64) (param $value i64) (result i32))
+```
+
+The behaviour of these functions is identical to their version 1 equivalent. Instead of allocating a buffer, writing `1` or `0` in it, and returning a pointer to it, the version 2 of these functions simply return 1 or 0.
+
+## Drawbacks
+
+- This RFC might be difficult to implement in Substrate due to the internal code design. It is not clear to the author of this RFC how difficult it would be.
+
+- In some situations, the runtime might still have to call `ext_allocator_malloc_version_1` then call one of the new functions, which is slightly less performant than when the two operations are combined into one call. The author of this RFC believes that this negligible, and that the performance saved by not allocating a buffer in most situations is worth the trade-off.
+
+## Prior Art
+
+The API of these new functions was heavily inspired by API used by the C programming language.
+
+## Unresolved Questions
+
+No unresolved questions.
+
+## Future Possibilities
+
+The existence of the host-side allocator has become questionable over time. It is implemented in a very naive way, and for determinism and backwards compatibility reasons it needs to be implemented exactly identically in every client implementation. Furthermore, runtimes make substantial use of heap memory allocations, and each allocation needs to go twice through the runtime <-> host boundary (once for allocating and once for freeing). Moving the allocator to the runtime side, while it would increase the size of the runtime, would be a good idea. But before the host-side allocator can be deprecated, all the host functions that make use of it need to be updated to not use it.
```

Comment:
How do you move the allocator to the runtime side?

Do you just give it a big block of memory and let it split it up as desired?

Reply:
I don't really see the problem? The runtime could just do whatever it wants with its memory.

If the question is how to make sure that the memory usage is bounded, my personal opinion would be that at initialization the runtime does `storage_get(":heappages")` and puts the value in its memory allocator, and the allocator crashes it it reaches the limit.

Alternatively, the wasm memory needs to be grown by the wasm code with a `memory.grow` opcode, so the host can detect when memory usage is over a certain threshold.

Reply:
Not an issue, I'm just not familiar enough with Wasm and was curious about how this would be done in practice

Reply:
Moving the allocator back into the runtime would make the runtime more similar to a normal program. The fact that the allocator is outside of the runtime is what requires additional code/hacks. So the answer on how to do move the allocator back to the runtime is simply "remove the hacks".